sky/skylet/providers/ibm/vpc_provider.py

```diff
@@ -0,0 +1,646 @@
+                if e.code == 404:
+                    print("VPC doesn't exist.") 
+                    deleting_resource = False
+                if e.code == 409:
+                    print("VPC still in use.")
+                    # will retry until cloud functions timeout. 
+                    time.sleep(5) 
+
+    def delete_vpc(vpc_id):
+        vpc_data = get_vpc_data(vpc_id)
+        if not vpc_data:
+            print((f"Failed to find a VPC with id={vpc_id}"))
+            return
+        print(f"Deleting vpc:{vpc_data['name']} with id:{vpc_id}")
+        delete_instances(vpc_data['id'])
+        delete_subnets(vpc_data)
+        delete_gateways(vpc_id)
+        delete_unbound_vpc(vpc_id)
+        print(f"VPC {vpc_data['name']} and its attached resources were deleted successfully")
+
+
+    def main(dict):
+        global ibm_vpc_client, RESOURCE_GROUP_ID, ibm_cloud_sdk_core, ibm_vpc
+        def install_package(package):
+            pip_location_stdout = subprocess.run(['which', 'pip'], capture_output=True, text=True)
+            pip_location = pip_location_stdout.stdout.strip()
+            subprocess.call([pip_location, 'install', package], stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+                
+        for package in ["ibm-vpc", "ibm-cloud-sdk-core"]:
+            install_package(package)
+
+        import ibm_vpc as _ibm_vpc
+        import ibm_cloud_sdk_core as _ibm_cloud_sdk_core
+        ibm_vpc = _ibm_vpc
+        ibm_cloud_sdk_core = _ibm_cloud_sdk_core
+
+        iam_api_key, RESOURCE_GROUP_ID, vpc_id, region = dict['iam_api_key'], dict['resource_group_id'], dict['vpc_id'], dict['region']
+
+        authenticator = ibm_cloud_sdk_core.authenticators.IAMAuthenticator(iam_api_key, url=None)
+        ibm_vpc_client = ibm_vpc.VpcV1('2022-06-30',authenticator=authenticator)
+
+        if not region:
+            raise Exception("VPC not found in any region")
+
+        ibm_vpc_client.set_service_url(f'https://{region}.iaas.cloud.ibm.com/v1')
+        
+        delete_vpc(vpc_id=vpc_id)
+        return {"Status": "Success"}
+        """)
+
+    def get_headers(self):
+        return {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+            "Authorization": "Bearer " + ibm.get_oauth_token()
+        }
+
+    def create_or_fetch_namespace(self):
+        """returns namespace id.
+        creates a namespace with given name in specified region. 
+        if namespace exists returns its id instead. """
+
+        def _create_new_namespace():
+            logger.info(f"Creating a new namespace: {self.namespace_name} in {self.namespace_region}")
+           
+            data = {"name": self.namespace_name, "resource_group_id": self.resource_group_id,
+                "resource_plan_id": "functions-base-plan"}
+        
+            res = requests.post(f'https://{self.namespace_region}.functions.cloud.ibm.com/api/v1/namespaces',
```

Comment:
Nit: the string `https://{self.namespace_region}.functions.cloud.ibm.com/api/v1/namespaces/` gets repeated a lot in the file. Maybe put it as a class or static variable?

Reply:
Agreed. Will be added as a class variable.